### PR TITLE
[refactor][vectorized] refactor first/last value agg functions

### PR DESCRIPTION
**Pull request by zhangstar333**
[refactor][vectorized] refactor first/last value agg functions

At the beginning, the implementation of **first/last value** is used for window functions, and **lead/lag** is also mixed in the implementation;
Later, when aiming at the **replace** model, also borrowed the basic implementation of first/last and made some changes.
Then, when **load**, also made some changes, so it became more and more complex. 
And now, the two are separated and further optimized respectively

# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/aggregate_functions/aggregate_function_reader_first_last.h [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+private:
+    std::string _copied_value;
+};
+
+template <typename ColVecType, bool result_is_nullable, bool arg_is_nullable, bool is_copy>
+struct FirstAndLastData {
+public:
+    using StoreType = std::conditional_t<is_copy, CopiedValue<ColVecType, result_is_nullable>,
+                                         Value<ColVecType, result_is_nullable>>;
+    static constexpr bool nullable = arg_is_nullable;
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_reader_first_last.h:78]: the code seems unless ?

File: be/src/vec/aggregate_functions/aggregate_function_reader_first_last.h [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+};
+
+template <typename Data>
+struct ReaderFunctionFirstNonNullData : Data {
+    void add(int64_t row, const IColumn** columns) {
+        if (this->has_set_value()) {
+            return;
+        }
+        if constexpr (Data::nullable) {
+            this->set_null_if_need();
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_reader_first_last.h:144]: really code will be here?

File: be/src/vec/aggregate_functions/aggregate_function_reader_first_last.h [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    void insert_result_into(ConstAggregateDataPtr place, IColumn& to) const override {
+        this->data(place).insert_result_into(to);
+    }
+
+    void add(AggregateDataPtr place, const IColumn** columns, size_t row_num,
+             Arena* arena) const override {
+        this->data(place).add(row_num, columns);
+    }
+
+    void reset(AggregateDataPtr place) const override {
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_reader_first_last.h:198]: code of reader should support reset

File: be/src/vec/aggregate_functions/aggregate_function_reader_first_last.h [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+            make_bool_variant(result_is_nullable), make_bool_variant(arg_is_nullable));
+    if (!res) {
+        LOG(WARNING) << " failed in  create_aggregate_function_" << name
+                     << " and type is: " << argument_types[0]->get_name();
+    }
+    return res;
+}
+
+template <bool is_copy>
+AggregateFunctionPtr create_aggregate_function_first_non_null_value(const std::string& name,
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_reader_first_last.h:264]: maybe should use #define to simplily the code

**Review by HappenLee — APPROVED** [review]

**Review by HappenLee — DISMISSED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**